File: 13212/inver.py
import sys

inversions = 0

def merge(l, r):
    global inversions
    temp = []

    # print(l, r)
    a = len(l)
    b = len(r)

    i = 0
    j = 0
    for k in range(a + b):
        if i >= a:
            temp.append(r[j])
            j = j + 1
            # print("a - i", a, i, a-i)
            inversions = inversions + (a - i)
        elif j >= b:
            temp.append(l[i])
            i = i + 1
        else:
            if l[i] <= r[j]:
                temp.append(l[i])
                i = i + 1
            elif l[i] > r[j]:
                temp.append(r[j])
                j = j + 1
                # print("a - i", a, i, a-i)
                inversions = inversions + (a - i)

    # print("inv:", inversions)
    return temp

def merge_sort(A):
    l = len(A)

    if l <= 1:
        return A
    else:
        mid = l // 2
        l = A[:mid]
        r = A[mid:]

        # print("l, r:", l, r)
        left = merge_sort(l)
        right = merge_sort(r)
        res = merge(left, right)

        # print("res: ", res)
        return res

if __name__ == '__main__':
    while True:
        n = int(sys.stdin.readline())

        if n == 0:
            break

        inversions = 0
        A = []
        for i in range(n):
            t = int(sys.stdin.readline())
            A.append(t)

        merge_sort(A)

        print(inversions)

File: 13212/inver.c
#include <stdio.h>

long long inversions = 0;
int n;

void merge(int A[], int p, int q, int r)
{
  int i, j, k;

  int n1 = q - p + 1;
  int n2 = r - q;

  int L[n1];
  int R[n2];

  for(i=0; i < n1; i++)
  {
    L[i] = A[p+i];
  }

  for(j=0; j < n2; j++)
  {
    R[j] = A[q+j+1];
  }

  i = 0;
  j = 0;

  for(k=p; k<=r; k++)
  {
    if(i >= n1)
    {
      A[k] = R[j];
      j++;
      inversions = inversions + (n1 - i);
    }
    else if(j >= n2)
    {
      A[k] = L[i];
      i++;
    }
    else
    {
      if(L[i] <= R[j])
      {
        A[k] = L[i];
        i++;
      }
      else
      {
        A[k] = R[j];
        j++;
        inversions = inversions + (n1 - i);
      }
    }
  }
}

void merge_sort(int A[], int p, int r)
{
  if (p < r)
  {
    int q = (p+r) / 2;

    merge_sort(A, p, q);
    merge_sort(A, q+1, r);
    merge(A, p, q, r);
  }
}

int main(int argc, char *argv[])
{
  int i;
  int A[1000005];

  while(1)
  {
    scanf("%d", &n);

    if(n == 0)
      break;

    for(i=0; i<n; i++)
    {
      scanf("%d", &A[i]);
    }

    inversions = 0;

    merge_sort(A, 0, n-1);

    printf("%lld\n", inversions);
  }

  return 0;
}
